### ISSUE TITLE
array slices: slice becomes invalid after appending to original array

### ISSUE THREAD

**Issue by UweKrueger** [vlang/v#8149]
**V version:** V 0.2.1 53941c4
**OS:** Ubuntu 20.04.1 LTS

**What did you do?**
```v
import os

fn main() {
    mut orig := [1, 2, 3, 4]
    slice := orig[1..3]
    println(slice)
    if os.args.len < 2 || os.args[1] != 'noloop' {
        for _ in 0 .. 1000 {
            orig << 9
        }
    }
    orig[2] = 7
    println(slice)
}
```

**What did you expect to see?**
```
[2, 3]
[2, 7]
 ```
**What did you see instead?**
```
[2, 3]
[0, 7938064]
```
However when called with argument `noloop` I get the expected result.

**Analysis**
Appending to the original causes a `realloc()` call for the array data buffer which possibly changes the memory location. The `orig` data pointer gets updated, the `slice` data pointer not.
There are many other issues concerning slices as they are implemented as now. E.g.  #6569, #7508, #119 and possibly others. I see several possible solutions:
 

1. @ntrel's suggestion from #6569 to store a pointer to the beginning of the original data in the slice. However, this has to be slightly modified regarding the issue here (#8149): Store the address of the original data pointer and its `cap` in the sliced array. So the original array's data pointer (and cap) get updated with every `realloc()` and the slices can see this (or there selves update the original array's data pointer and cap).
The original array would have a kind of flag to be identified at run time.
With this approach `slices` would be internally (at run time) different to `arrays` but not from the user's point of view and not for the compiler.
2. @emily33901 has suggested to make slices fixed sized views to parts of arrays. This would avoid problems like the one reported in #6569, too. In this case, too, the slice must have a reference to the data pointer of the original array to fix the issue here (#8149).
Slices would be different from arrays for the user and the compiler - you could append to arrays - but not to slices. Functions would usually be written for slices and the compiler would coerce a slice automatically when called with an array. A syntax for such a function might be: `fn f(a [..]int) {...}` with `..` between the brackets indicating that there is a fixed size.
This suggestion is the closest to what `Go` does - Go's slices however *can* be changed in size, but this causes a lot of problems. We must also consider that with this approach we would actually have 3 different array-like types in V: variable sized arrays, fixed size arrays (on the stack) and slices.
3. Make slice operations always clone the relevant part or the array. This is how Python does it (and this would by *my* (@UweKrueger) favorite). Implementing this would be easy: just replace every call to `slice()` in the generated C code with a call to `slice_clone()`.
With this approach a slice would be a fully featured independent copy and users wouldn't have to think about the array they were made from. However, there is no aliasing (which might be good or bad). Functions that are supposed to modify only a small part of a large array would have to be called with this array and the index range - in the other suggestions a slice could be an abstraction of a reference to a part of the array.

**Comment by ntrel** [comment]
> Store the address of the original data pointer and its cap in the sliced array. So the original array's data pointer (and cap) get updated with every realloc() **and the slices can see this** (or there selves update the original array's data pointer and cap).
The original array would have a kind of flag to be identified at run time.

(My way `slice` and `orig` are the same type). I would use a reference count (stored in allocated memory just before the elements). Slicing is very common for strings and arrays so reference counting is a good strategy (except where the compiler can prove it doesn't escape the original owner variable). Allocating copies is expensive, this avoids that. When memory is exhausted, `orig` would only use `realloc` if the reference count is 1 and `orig.start == orig.data`.

**Comment by UweKrueger** [comment]
> (My way `slice` and `orig` are the same type). I would use a reference count (stored in allocated memory just before the elements)...
> _Update_: So `slice` is never changed by appending to `orig`.

The problem is what happens when by appending to `slice` or `orig` the data doesn't fit anymore in the memory space. `realloc()` will move the data. So you need something like this:
```
orig -----> {ref_count, data_ptr } -----> [data...]
             ^ ^
slice -------+ |
               |
slice2 --------+
```
So both slices and orig point to a (not moving) control structure that contains the ref-counter and the data pointer. But you're right: orig and slice were indistinguishable.

**Comment by UweKrueger** [comment]
> _Update_: So `slice` is never changed by appending to `orig`.

@ntrel This is not true IMHO. If you add elements to the slice and then add elements to orig, the elements added to orig overwrite the elements in slice. This is exactly the confusing situation that exists in Go.

**Comment by spytheman** [comment]
Using another level of indirection like that, will solve several problems, but it may introduce a potential performance problem.
Every array access will need to go through that centralized storage descriptor. We would have to measure what would be the impact 🤔 .

**Comment by ntrel** [comment]
> both slices and orig point to a (not moving) control structure

That is not my idea. I explained what happens, this is true when appending to `slice` too:

> When memory is exhausted, orig would only use realloc if the reference count is 1 and orig.start == orig.data.

Otherwise, `orig` allocates new memory and copies existing elements.

```
                  {cap, ref_count, elem0, elem1....}
orig.data -------------------------^
orig.start ------------------------^
slice.data ------------------------^
slice.start ------------------------------^
```
slice can find `cap` and `ref_count` by looking behind slice.start.

**Comment by spytheman** [comment]
what happens if I call `a.delete(0)` ?

**Comment by ntrel** [comment]
> If you add elements to the slice and then add elements to orig, the elements added to orig overwrite the elements in slice

You're right. Some other mechanism can probably be used to force a copy instead.

**Comment by UweKrueger** [comment]
> Otherwise, `orig` allocates new memory and copies existing elements.

And what happens with the old orig? The slices are still pointing there so it can't be deleted. But even if the old data of orig is kept the slices would not refer to orig any more which results in undefined behaviour (since you never know when `realloc()` is invoked.

**Comment by ntrel** [comment]
> The slices are still pointing there so it can't be deleted

If the elements are being copied, of course the original can't be deleted.

> the slices would not refer to orig any more which results in undefined behaviour (since you never know when realloc() is invoked

Huh, realloc would not be invoked if a valid slice exists. (When the slice goes out of scope, the ref_count is decremented).

**Comment by ntrel** [comment]
> what happens if I call a.delete(0) ?

That would affect the elements in the slice too.

**Comment by UweKrueger** [comment]
> Huh, realloc would not be invoked if a valid slice exists. (When the slice goes out of scope, the ref_count is decremented).

@ntrel Please forget about reference counting and memory management for a moment (these things are no problem). I'm talking about the question if slices and orig always having the same corresponding elements (not equal elements) - and changes in one of them affects the other. If slices are designed as references this should always be the case. But if the slices point to the old orig and the new orig has moved its data (due to necessary reallocation) the connection is lost.
This is why I think a non-moving data structure is needed in between.

**Comment by ntrel** [comment]
> If slices are designed as references this should always be the case

My solution does not distinguish between a slice and an array, so it does not allow slice elements to mirror what elements the original has if the original has been appended to. Before appending they refer to the same elements so a change in either the slice or the original will be seen in the other one.

This is what D does. It is better than Go because appending can be made to work safely. The programmer would have to be aware of how appending affects element sharing, or avoid slices or use immutable arrays.

**Comment by UweKrueger** [comment]
@ntrel I think I now get it :-)

> ... so it does not allow slice elements to mirror what elements the original has if the original has been appended to. Before appending (at least where `cap is exhausted) they refer to the same elements so a change in either the slice or the original will be seen in the other one.

I'd suggest a copy should *always* be made when appending to an array/slice and there is another array/slice referring the same data - even if `cap` is not exhausted. This would then also solve [this](https://github.com/vlang/v/issues/8149#issuecomment-761545654) problem and otherwise it would be hard to predict what happens. (The user should not deal with `cap` - let alone if the heap space is big enough to `realloc()` without movement.)
So the user would know for sure that after appending an element the connection to any other slices is lost. 

> This is what D does. It is better than Go because appending can be made to work safely. The programmer would have to be aware of how appending affects element sharing, or avoid slices or use immutable arrays.

Agreed, this is indeed better than what Go does - and it avoids introducing a new data type for `slice`. It's a solution I could live with.

**Comment by ntrel** [comment]
> I'd suggest a copy should always be made when appending to an array/slice and there is another array/slice referring the same data 

Sounds good. It's sub-optimal in the case where a slice is likely after each append, but I would guess that's not the common case. We could even have a special function for that case (if needed), `fn (mut a array) append_reuse(ElemType)` that always reuses any available space, even if there are shared owners.

There are other ways of preventing two shared owners both appending, but the advantage of this approach is that it's easier for the programmer to know if a slice still aliases the same memory as the original array.

**Comment by emily33901** [comment]
> I'd suggest a copy should _always_ be made when appending to an array/slice and there is another array/slice referring the same data

Who owns the original memory after the clone is made? Who is responsible for freeing it?

**Comment by UweKrueger** [comment]
> Who owns the original memory after the clone is made? Who is responsible for freeing it?

This is not really a problem. It can be done by reference counting. The array that creates the copy gets ref_count=1 for this copy and the ref_count for the original is decreased. Memory is freed when ref_count becomes 0.

**Comment by penguindark** [comment]
other test case from closed #3352 :
```v
fn main() {
        mut a := [1,2,3]
        a = a[1..] // `a.delete(0)` or `a = a[..2]` works fine
        a << 4
}
```
